plugins/dead_code_elimination/dead_code_elimination_main.ml

```diff
@@ -23,95 +38,155 @@ end
+let live_phi dead blk = Term.filter phi_t ~f:(is_alive dead) blk
+
+let live_def checked dead blk =
+  Term.filter def_t blk ~f:(fun d ->
+      if checked (Def.lhs d) then is_alive dead d
+      else true)
+
+let substitute sub vars =
+  (object
+    inherit Term.mapper as super
+    method! map_var v =
+      match Map.find vars v with
+      | None -> Bil.var v
+      | Some e -> e
+    method! map_exp e = Exp.fold_consts (super#map_exp e)
+  end)#map_sub sub
+
+(* A simple constant propagation. Note, the input is required to be in SSA. *)
+let propagate_consts checked sub =
+  let vars = (object
+    inherit [exp Var.Map.t] Term.visitor
+    method! enter_def t vars =
+      let v = Def.lhs t in
+      if checked v then
+        match Def.rhs t with
+        | Bil.Unknown _ | Bil.Int _ as exp ->
+          Map.add vars ~key:v ~data:exp
+        | _ -> vars
+      else vars
   end)#visit_sub sub Var.Map.empty in
-  Term.map blk_t sub ~f:(Blk.map_exp ~f:(fun exp ->
-      Map.fold vars ~init:exp ~f:(fun ~key:pat ~data:rep ->
-          Exp.substitute Bil.(var pat) rep)))
-
-let rec process proj =
-  let prog = Project.program proj in
-  report_progress ~note:"ssa" ();
-  let subs = Term.enum sub_t prog |> Seq.map ~f:Sub.ssa in
+  substitute sub vars
+
+let clean checked dead sub =
+  Term.map blk_t sub ~f:(fun b -> live_def checked dead b |> live_phi dead)
+
+let return_from_ssa sub =
+  let drop_index t = (object
+    inherit Term.mapper
+    method! map_sym var = Var.base var
+  end)#map_sub t in
+  Term.map blk_t sub ~f:(Term.filter phi_t ~f:(fun _ -> false)) |>
+  drop_index
+
+let is_flag arch =
+  let module T = (val (target_of_arch arch)) in
+  T.CPU.is_flag
+
+let free_vars prog =
   report_progress ~note:"free-vars" ();
-  let free = union subs ~init:Var.Set.empty ~f:free_vars in
-  report_progress ~note:"dead-vars" ();
-  let dead = union subs ~init:Tid.Set.empty ~f:(compute_dead free) in
-  let live t = not (Set.mem dead (Term.tid t)) in
-  report_progress ~note:"clean" ();
-  let clean sub = Term.map blk_t sub ~f:(Term.filter def_t ~f:live) |>
-                  propagate_consts in
-  report_progress ~note:"updating" ();
-  if Set.is_empty dead then proj
-  else
-    Term.map sub_t prog ~f:clean |>
-    Project.with_program proj |>
-    process
-
-let () = Config.when_ready (fun _ ->
-    Project.register_pass ~deps:["api"] ~autorun:true process)
-;;
-
-Config.manpage [
-  `S "SYNOPSIS";
-  `Pre "
-    $(b,--no-$mname)
+  Term.enum sub_t prog |>
+  Seq.fold ~init:Var.Set.empty ~f:(fun free sub ->
+      let free = Set.union free (Sub.free_vars sub) in
+      Seq.fold ~init:free (Term.enum arg_t sub)
+        ~f:(fun free arg ->
+            Set.union free (Exp.free_vars (Arg.rhs arg))))
+
+let union ~init ~f = List.fold ~init ~f:(fun xs x -> Set.union xs (f x))
+
+let process arch prog levels =
+  let checked = check_level (is_flag arch) levels in
+  let free = free_vars prog in
+  let rec run subs =
```

Comment:
apply optimization on each subroutine independently (except the protected set which should be computed interprocedurally). Then store the optimization for each subroutine in the cache (do not forget to include the digest of the protected set into the digest of a subroutine). 

In general always think about Google Chrome. The current implementation will require you to load in the memory and store in a list 3 google chromes if not more. While we can't even allow to store one. 

Basically, try to implement you analysis in a such way that it keeps in memory as small data as possible, ideally, no more than one subroutine. Also try to drop unused data as soon as possible, i.e., once you have a subroutine ready push it back to the project, store the cache, and keep going with the next one).